python_modules/libraries/dagster-dg/dagster_dg/utils/__init__.py

```diff
@@ -440,64 +459,170 @@ def parse_json_option(context: click.Context, param: click.Option, value: str):
+
 
-def get_toml_value(
-    doc: tomlkit.TOMLDocument,
-    path: Iterable[str],
+def get_toml_node(
+    doc: TomlDoc,
+    path: TomlPath,
     expected_type: Union[type[T], tuple[type[T], ...]],
 ) -> T:
     """Given a tomlkit-parsed document/table (`doc`),retrieve the nested value at `path` and ensure
     it is of type `expected_type`. Returns the value if so, or raises a KeyError / TypeError if not.
     """
-    current: Any = doc
-    for key in path:
-        # If current is not a table/dict or doesn't have the key, error out
-        if not isinstance(current, dict) or key not in current:
-            raise KeyError(f"Key '{key}' not found in path: {'.'.join(path)}")
-        current = current[key]
-
-    # Finally, ensure the found value is of the expected type
-    if not isinstance(current, expected_type):
+    value = _gather_toml_nodes(doc, path)[-1]
+    if not isinstance(value, expected_type):
         expected_types = expected_type if isinstance(expected_type, tuple) else (expected_type,)
         type_str = " or ".join(t.__name__ for t in expected_types)
         raise TypeError(
-            f"Expected '{'.'.join(path)}' to be {type_str}, "
-            f"but got {type(current).__name__} instead."
+            f"Expected '{toml_path_to_str(path)}' to be {type_str}, "
+            f"but got {type(value).__name__} instead."
         )
-    return current
+    return value
 
 
-def has_toml_value(doc: tomlkit.TOMLDocument, path: Sequence[str]) -> bool:
+def has_toml_node(doc: TomlDoc, path: TomlPath) -> bool:
     """Given a tomlkit-parsed document/table (`doc`), return whether a value is defined at `path`."""
-    leading_path, key = path[:-1], path[-1]
-    current = doc
-    for leading_key in leading_path:
-        if not isinstance(current, dict) or leading_key not in current:
-            return False
-        current = current[leading_key]
-    return isinstance(current, dict) and key in current
+    try:
+        _gather_toml_nodes(doc, path)
+    except KeyError:
+        return False
+    return True
 
 
-def delete_toml_value(doc: tomlkit.TOMLDocument, path: Sequence[str]) -> None:
+def delete_toml_node(doc: TomlDoc, path: TomlPath) -> None:
     """Given a tomlkit-parsed document/table (`doc`), delete the nested value at `path`. Raises
-    an error if the leading keys do not already lead to a dictionary.
+    an error if the leading keys do not already lead to a TOML container node.
     """
-    dct = get_toml_value(doc, path[:-1], dict) if len(path) > 1 else doc
-    del dct[path[-1]]
+    nodes = _gather_toml_nodes(doc, path)
+    container = nodes[-2]
+    key_or_index = path[-1]
+    if isinstance(container, dict):
+        del container[path[-1]]
+    elif isinstance(container, list):
+        assert isinstance(key_or_index, int)  # We already know this from _traverse_toml_path
+        container.pop(key_or_index)
+    else:
+        raise Exception("Unreachable.")
```

Comment:
wow crazy that we can't even use check here

Reply:
Yep when we factor out a common utils package we'll have access to check.